drivers/usb/common/nrf_usbd_common/nrf_usbd_common.c

```diff
@@ -1494,72 +1217,74 @@ static void usbd_enable(void)
 }
 /** @} */
 
-/**
- * @brief USBD interrupt service routines.
- *
- */
-static const nrfx_irq_handler_t m_isr[] = {[USBD_INTEN_USBRESET_Pos] = ev_usbreset_handler,
-					   [USBD_INTEN_STARTED_Pos] = ev_started_handler,
-					   [USBD_INTEN_ENDEPIN0_Pos] = ev_dma_epin0_handler,
-					   [USBD_INTEN_ENDEPIN1_Pos] = ev_dma_epin1_handler,
-					   [USBD_INTEN_ENDEPIN2_Pos] = ev_dma_epin2_handler,
-					   [USBD_INTEN_ENDEPIN3_Pos] = ev_dma_epin3_handler,
-					   [USBD_INTEN_ENDEPIN4_Pos] = ev_dma_epin4_handler,
-					   [USBD_INTEN_ENDEPIN5_Pos] = ev_dma_epin5_handler,
-					   [USBD_INTEN_ENDEPIN6_Pos] = ev_dma_epin6_handler,
-					   [USBD_INTEN_ENDEPIN7_Pos] = ev_dma_epin7_handler,
-					   [USBD_INTEN_EP0DATADONE_Pos] = ev_setup_data_handler,
-					   [USBD_INTEN_ENDISOIN_Pos] = ev_dma_epin8_handler,
-					   [USBD_INTEN_ENDEPOUT0_Pos] = ev_dma_epout0_handler,
-					   [USBD_INTEN_ENDEPOUT1_Pos] = ev_dma_epout1_handler,
-					   [USBD_INTEN_ENDEPOUT2_Pos] = ev_dma_epout2_handler,
-					   [USBD_INTEN_ENDEPOUT3_Pos] = ev_dma_epout3_handler,
-					   [USBD_INTEN_ENDEPOUT4_Pos] = ev_dma_epout4_handler,
-					   [USBD_INTEN_ENDEPOUT5_Pos] = ev_dma_epout5_handler,
-					   [USBD_INTEN_ENDEPOUT6_Pos] = ev_dma_epout6_handler,
-					   [USBD_INTEN_ENDEPOUT7_Pos] = ev_dma_epout7_handler,
-					   [USBD_INTEN_ENDISOOUT_Pos] = ev_dma_epout8_handler,
-					   [USBD_INTEN_SOF_Pos] = ev_sof_handler,
-					   [USBD_INTEN_USBEVENT_Pos] = ev_usbevent_handler,
-					   [USBD_INTEN_EP0SETUP_Pos] = ev_setup_handler,
-					   [USBD_INTEN_EPDATA_Pos] = ev_epdata_handler};
-
 /**
  * @name Interrupt handlers
  *
  * @{
  */
 void nrf_usbd_common_irq_handler(void)
 {
-	const uint32_t enabled = nrf_usbd_int_enable_get(NRF_USBD);
-	uint32_t to_process = enabled & ~NRF_USBD_INT_EP0SETUP_MASK;
-	uint32_t active = 0;
+	uint32_t epdatastatus = 0;
 
-	/* Check all enabled interrupts */
-	while (to_process) {
-		uint8_t event_nr = NRF_CTZ(to_process);
+	/* Clear EPDATA event and only then get and clear EPDATASTATUS to make
+	 * sure we don't miss any event.
+	 */
+	if (NRF_USBD->EVENTS_EPDATA) {
+		NRF_USBD->EVENTS_EPDATA = 0;
+		epdatastatus = NRF_USBD->EPDATASTATUS;
+		NRF_USBD->EPDATASTATUS = epdatastatus;
```

Comment:
In my opinion, the code is much easier to follow if it can be easily be matched against the Product Specification document. When we arrive at native driver I would like to not use any USBD HAL at all.

Reply:
There are certain advantages of using the nrf HAL, like the possibility of handling in that layer some hardware anomalies or differences in registers layout between SoC series. It also makes it easy to use nrfx drivers in simulation. I agree that in the case of USBD, those advantages might not be that evident at the moment, but in my opinion, we should stay consistent in using nrf HAL in nRF drivers. Also because this file already uses USBD HAL functions, I think we should keep consistency in this regard.

Reply:
What is evident is that USBD HAL is really complicating the understanding of what the driver does. In my view the "file already uses" is really "file still uses" because I see no point in changing something without a real cause.

If the real cause would be "be consistent regarding the use of USBD HAL", then my strongly preferred solution to that would be to not use USBD HAL at all.

Reply:
I don't quite understand why is that a problem to use the USBD HAL in those few added accesses to USBD registers. It's much less effort than removing all uses of that HAL. But I still think this file should be consistent in using or not using the USBD HAL, so please do what you prefer to achieve this.

Reply:
Insisting on being consistent in using or not using USBD HAL kinda defeats the "enables us to refactor the code towards a native driver in a step by step manner". As you noticed it would be definitely harder to remove all remaining uses of USBD HAL at this point and the removal itself at this point serves no rel purpose but to please the reviewer.

If you still insist, then fine, I'll remove all uses of USBD HAL but I don't think it is worth the hassle at this point. And at the same time sticking to using the USBD HAL in newly rewritten code just for [foolish consistency](https://peps.python.org/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds) is silly. If you insist then it'll be next commit and the commit message will include that it is done without purpose other than to please the reviewer.